### DISCUSSION TITLE
Pimcore Data Importer: Mapping Select Field Values from Excel with Options Provider

### DISCUSSION THREAD

**Discussion by Chauhan-Mukesh** [pimcore/pimcore#17089] [Q&A / Support]
Hi everyone,

I'm facing challenges mapping a select field named "colorway" in Pimcore's Data Importer. I'm importing data from an Excel file, where the "colorway" column contains the colorway name.

In Pimcore, the "colorway" select field is configured using an Options Provider ``` \App\Lib\OptionProviders\GeneralOptionsProvider``` with the following configuration:

```
{"class": "Colorway","key": "name","value": "id","extra": "","orderKey":"name","order":"asc"}
```
This means the select field expects the colorway object's ID as its value. However, the Data Importer's "Load DataObject" transformation pipeline returns the entire colorway object when I map the field.

How can I modify the mapping or the transformation pipeline so that the Data Importer correctly assigns the colorway object's ID to the select field?

Any help or suggestions would be greatly appreciated!

- **Pimcore version:** 11.2.3.3
- **Data Importer version:** 1.8
- **Excel file format:** xlsx

**Data Importer Configuration**
![7](https://github.com/pimcore/pimcore/assets/70336897/dce049f5-e9c8-4e32-830d-ce1e22182914)

**Class Configuration for _colorway_ field**
![9](https://github.com/pimcore/pimcore/assets/70336897/9f5e050d-9de7-4ea1-b68a-8b7f310f99ba)

**Comment by marko91** [comment]
@Chauhan-Mukesh having a similar problem, did you manage to solve it?

**Reply by Chauhan-Mukesh** [reply]
Hi @marko91,

No I didn't get any success on this issue, but I used an alternate approch to do so.

I created an event listener which gives me the row data while import and I used that data to set the value in required field via code only in listener.

Example:
```
<?php
/**
 * DataImportListener.php.php
 *
 * This file will listen to the data import event and set the brand id in product object
 * if brand is available in the system.
 *
 * @category   EventListener
 * @package    AppUiBundle\EventListener
 * @version    1.0.0
 *
 * @author     Mukesh Chauhan
 */
namespace AppUiBundle\EventListener;

use Pimcore\Bundle\DataImporterBundle\Event\DataObject\PreSaveEvent;
use Pimcore\Model\DataObject;

class DataImportListener
{
    public function onPreSave(PreSaveEvent $event): void
    {
        $rowData = $event->getRawData();
        $dataObject = $event->getDataObject();
        if ($dataObject instanceof DataObject\Product){
            $brand = isset($rowData[11]) && !empty($rowData[11])?$rowData[11]:null;
            if ($brand){
                $brandObject = DataObject\Brand::getByName($brand,1);
                if ($brandObject){
                    $dataObject->setBrand($brandObject->getId());
                }
            }
        }
    }
}

```

```
services:
    AppUiBundle\EventListener\DataImportListener:
        tags:
            - { name: kernel.event_listener, event: Pimcore\Bundle\DataImporterBundle\Event\DataObject\PreSaveEvent, method: onPreSave }
```
Although this is not a good approach when we have to deal with large number of fields of similar kind in other classes too, because it will be a kind of increasing the unnessesary code.

**Reply by marko91** [reply]
Thanks for the fast reply @Chauhan-Mukesh, I was able to solve my issue.

The CSV that we were importing had a string field for currency which was representing the currencyCode in the Currency DataObjects (for Currencies we have our own objects with name & currencyCode - the field type in the object that we were trying to import is Select with a custom OptionsProvider).

I was able to map the CSV currency field directly to the "Currency" Select field in the objects we were importing. I had to change the options provider to return the currencyCode for the value.


![image](https://github.com/pimcore/pimcore/assets/2420568/1c5e2f14-8851-4c23-9d05-84d983d433fc)
![image](https://github.com/pimcore/pimcore/assets/2420568/325b5056-8bc0-4471-866e-719f10e8192d)

The Source Attribute is a String representing the currencyCode - for instance "1"
The Data Target is a currency Select field.

I believe the same approach could work for you since it is a similar issue.

**Reply by Chauhan-Mukesh** [reply]
HI @marko91 

Yes the change you mentioned can be done, But in my case the option provider I am using a General Option Provider which can be used by any class, so this case does not suits better to me.
